drivers/adc/adc_demo/adc_demo.c

```diff
@@ -0,0 +1,126 @@
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include "adc_demo.h"
+#include "error.h"
+#include "util.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/***************************************************************************//**
+ * @brief init function for the adc demo driver
+ * @param desc - descriptor for the adc
+ * @param param - initialization param for adc
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adc_demo_init(struct adc_demo_desc **desc,
+		      struct adc_demo_init_param *param)
+{
+	struct adc_demo_desc *adesc;
+	adesc = (struct adc_demo_desc*)calloc(1, sizeof(*adesc));
+
+	if(!adesc)
+		return -ENOMEM;
+
+	*desc = adesc;
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * @brief free allocated resources
+ * @param desc - descriptor for the adc
+ * @return SUCCESS in case of success, FAILURE otherwise.
+*******************************************************************************/
+int32_t adc_demo_remove(struct adc_demo_desc *desc)
+{
+	if(!desc)
+		return -EINVAL;
+
+	free(desc);
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * @brief read function for the adc demo driver
+ * @param desc - descriptor for the adc
+ * @param reg_index - the address at which we want to read
+ * @param readval- the value read from register
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adc_demo_reg_read(struct adc_demo_desc *desc, uint8_t reg_index,
+			  uint8_t *readval)
+{
+
+	if(!desc || reg_index > ARRAY_SIZE(desc->reg))
```

Comment:
this still does not seem correct you fail only if reg_index > ARRAY_SIZE(desc->reg)

let's say desc->reg[16] isssss the array
ARRAY_SIZE(desc->reg) = 16
reg_index = 16 passes and you get to 

*readval =  desc->reg[reg_index]; // here you try to access reg[16] - out of bounds error

Reply:
changed it to >=